route/grafananet.go

```diff
@@ -143,130 +147,116 @@ func NewGrafanaNet(key, prefix, sub, regex, addr, apiKey, schemasFile string, sp
 		Transport: transport,
 	}
 
-	go r.run()
 	return r, nil
 }
 
-func (route *GrafanaNet) run() {
-	metrics := make([][]*schema.MetricData, route.concurrency)
-	route.wg.Add(route.concurrency)
-	for i := 0; i < route.concurrency; i++ {
-		metrics[i] = make([]*schema.MetricData, 0, route.flushMaxNum)
-
-		// start up our goroutines for writing data to tsdb-gw
-		go route.flush(i)
-	}
-
-	flush := func(shard int) {
-		if len(metrics[shard]) == 0 {
-			return
-		}
-		mda := schema.MetricDataArray(metrics[shard])
-		data, err := msg.CreateMsg(mda, 0, msg.FormatMetricDataArrayMsgp)
-		if err != nil {
-			panic(err)
-		}
-		route.writeQueues[shard] <- data
-		route.numOut.Inc(int64(len(metrics[shard])))
-		metrics[shard] = metrics[shard][:0]
-
-	}
-
-	hasher := fnv.New32a()
+// runs manages incoming and outgoing data for a shard
+func (route *GrafanaNet) run(in chan []byte) {
+	var metrics []*schema.MetricData
+	buffer := new(bytes.Buffer)
 
 	ticker := time.NewTicker(route.flushMaxWait)
 	for {
 		select {
-		case buf := <-route.buf:
+		case buf := <-in:
 			route.numBuffered.Dec(1)
 			md, err := parseMetric(buf, route.schemas, route.orgId)
 			if err != nil {
 				log.Error("RouteGrafanaNet: %s", err)
 				continue
 			}
 			md.SetId()
+			metrics = append(metrics, md)
 
-			//re-use our []byte slice to save an allocation.
-			buf = md.KeyBySeries(buf[:0])
-			hasher.Reset()
-			hasher.Write(buf)
-			shard := int(hasher.Sum32() % uint32(route.concurrency))
-			metrics[shard] = append(metrics[shard], md)
-			if len(metrics[shard]) == route.flushMaxNum {
-				flush(shard)
+			if len(metrics) == route.flushMaxNum {
+				metrics = route.retryFlush(metrics, buffer)
 			}
 		case <-ticker.C:
-			for shard := 0; shard < route.concurrency; shard++ {
-				flush(shard)
-			}
+			metrics = route.retryFlush(metrics, buffer)
```

Comment:
Right now, we could flush due to reaching FlushMaxNum then receive just 1 metric before the ticker fires we we flush again.
now that there is a single payload to flush, we should change this from a ticker to a timer that is reset when we flush.  
ie.
```
t := time.NewTimer(route.FlushMaxWait)
for {
  select {
  case buf := <- in:
    ...
    if len(metrics) == route.flushMaxNum {
      metrics = route.retryFlush(metrics, buffer)
      // reset our timer
      if !t.Stop() {
	<-t.C
      }
      t.Reset(route.FlushMaxWait)
    }
  case <timer.C:
    metrics = route.retryFlush(metrics, buffer)
    t.Reset(route.FlushMaxWait)
...
```